src/main/java/mate/academy/bookstore/service/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,83 @@
+package mate.academy.bookstore.service;
+
+import lombok.RequiredArgsConstructor;
+import mate.academy.bookstore.exception.EntityNotFoundException;
+import mate.academy.bookstore.mapper.CartItemMapper;
+import mate.academy.bookstore.mapper.ShoppingCartMapper;
+import mate.academy.bookstore.model.CartItem;
+import mate.academy.bookstore.model.ShoppingCart;
+import mate.academy.bookstore.model.dto.shopingcart.CartItemRequestDto;
+import mate.academy.bookstore.model.dto.shopingcart.ShoppingCartResponseDto;
+import mate.academy.bookstore.repository.shoppingcart.CartItemRepository;
+import mate.academy.bookstore.repository.shoppingcart.ShoppingCartRepository;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final CartItemMapper cartItemMapper;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemRepository cartItemRepository;
+
+    @Override
+    public ShoppingCartResponseDto addCartItem(CartItemRequestDto requestDto, Long userId) {
+        boolean isPresent = false;
+        ShoppingCart shoppingCartByUser
+                = shoppingCartRepository.findShoppingCartByUserId(userId);
+        for (CartItem item : shoppingCartByUser.getCartItems()) {
+            if (item.getBook().getId() == requestDto.getBookId()) {
+                isPresent = true;
+                break;
+            }
+        }
+        CartItem cartItem = cartItemMapper.mapToModel(requestDto);
+        if (!isPresent) {
+            shoppingCartByUser.getCartItems().add(cartItemRepository.save(cartItem));
+        } else {
+            cartItem.setQuantity(requestDto.getQuantity());
+            cartItemRepository.save(cartItem);
+        }
+        shoppingCartRepository.save(shoppingCartByUser);
+        return shoppingCartMapper.mapToDto(shoppingCartByUser);
```

Comment:
```suggestion
       User user = userRepository.findById(userId)
      .orElseThrow(EntityNotFoundException.supplier("User not found"));
    Book book = bookRepository.findById(request.bookId())
      .orElseThrow(EntityNotFoundException.supplier("Book not found"));

    ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId).orElseGet(() -> {
      ShoppingCart newShoppingCart = new ShoppingCart();
      newShoppingCart.setUser(user);
      shoppingCartRepository.save(newShoppingCart);
      return newShoppingCart;
    });

    CartItem cartItem = new CartItem();
    cartItem.setBook(book);
    cartItem.setShoppingCart(shoppingCart);
    cartItem.setQuantity(request.quantity());
    cartItemRepository.save(cartItem);

    return shoppingCartMapper.toDto(shoppingCart);
```

this way it is just easier to undersant